File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    bool contains(TreeNode* root , TreeNode* p){
        if(!root) return false;
        if(root->val == p->val) return true;
        if(root->val > p->val){
            return contains(root->left,p);
        }
        else return contains(root->right,p);
    }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root) return nullptr;
        TreeNode* lca = nullptr;
        if(contains(root,p) and contains(root,q)){
            lca =  root;
        };
        if(root->val > p->val and root->val > q->val){
            lca =  lowestCommonAncestor(root->left,p,q);
        }
        else if(root->val < p->val and root->val < q->val) {
            lca =  lowestCommonAncestor(root->right,p,q);
        }
        return lca;
    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {

    private (int wr, int wor) CountMax(TreeNode root){
        if(root == null){
            return (0,0);
        }
        (int wr, int wor) lrs = CountMax(root.left);
        (int wr, int wor) rrs = CountMax(root.right);

        return (root.val + lrs.wor + rrs.wor, Math.Max(lrs.wr,lrs.wor) + Math.Max(rrs.wr,rrs.wor));
    }
    public int Rob(TreeNode root) {
         (int wr, int wor) rs = CountMax(root);
         return Math.Max(rs.wr,rs.wor);
    }
}